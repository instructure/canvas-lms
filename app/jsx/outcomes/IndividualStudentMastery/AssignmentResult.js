/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import I18n from 'i18n!outcomes'
import _ from 'lodash'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import View from '@instructure/ui-layout/lib/components/View'
import Link from '@instructure/ui-elements/lib/components/Link'
import Text from '@instructure/ui-elements/lib/components/Text'
import IconAssignment from '@instructure/ui-icons/lib/Line/IconAssignment'
import IconQuiz from '@instructure/ui-icons/lib/Line/IconQuiz'
import * as shapes from './shapes'
import Ratings from '../../rubrics/Ratings'

const AssignmentResult = ({ outcome, result, outcomeProficiency }) => {
  const { ratings } = outcome
  const { html_url: url, name, submission_types: types } = result.assignment
  const isQuiz = types && types.indexOf('online_quiz') >= 0

  return (
    <Flex padding="small" direction="column" alignItems="stretch">
      <FlexItem>
        <Link href={ url }>
          <Flex alignItems="center">
            <FlexItem><Text size="medium">
              {
                isQuiz ? <IconQuiz /> : <IconAssignment />
              }
            </Text></FlexItem>
            <FlexItem padding="0 x-small"><Text weight="bold">{ name }</Text></FlexItem>
          </Flex>
        </Link>
      </FlexItem>
      <FlexItem padding="x-small 0">
        <View padding="x-small 0 0 0">
          <Text size="small" fontStyle="italic" weight="bold">{
            result.hide_points ? I18n.t('Your score') : I18n.t('Your score: %{score}', { score: result.score })
          }</Text>
        </View>
      </FlexItem>
      <FlexItem borderWidth="small">
        <div className="react-rubric">
          <div className="ratings">
            <Ratings
              tiers={ratings}
              points={result.score}
              hidePoints={result.hide_points}
              useRange={false}
              customRatings={_.get(outcomeProficiency, 'ratings')}
              defaultMasteryThreshold={outcome.mastery_points}
              pointsPossible={outcome.points_possible}
              assessing={false}
              isSummary={false}
            />
          </div>
        </div>
      </FlexItem>
    </Flex>
  )
}

AssignmentResult.propTypes = {
  result: shapes.outcomeResultShape.isRequired,
  outcome: shapes.outcomeShape.isRequired,
  outcomeProficiency: shapes.outcomeProficiencyShape
}

AssignmentResult.defaultProps = {
  outcomeProficiency: null
}

export default AssignmentResult
