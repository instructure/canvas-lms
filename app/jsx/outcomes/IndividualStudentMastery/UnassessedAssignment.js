/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {object} from 'prop-types'
import I18n from 'i18n!IndividualStudentMasteryUnassessedAssignment'
import _ from 'lodash'
import {ApplyTheme} from '@instructure/ui-themeable'
import {View} from '@instructure/ui-layout'
import {Button} from '@instructure/ui-buttons'
import {IconAssignmentLine, IconQuizLine} from '@instructure/ui-icons'
import {List} from '@instructure/ui-elements'

const UnassessedAssignment = ({assignment}) => {
  const {id, url, submission_types, title} = assignment
  return (
    <List.Item key={id}>
      <View padding="small" display="block">
        <ApplyTheme theme={{[Button.theme]: {linkColor: '#68777D', fontWeight: '700'}}}>
          <Button
            href={url}
            variant="link"
            theme={{mediumPadding: '0', mediumHeight: 'normal'}}
            icon={_.includes(submission_types, 'online_quiz') ? IconQuizLine : IconAssignmentLine}
          >
            {title} ({I18n.t('Not yet assessed')})
          </Button>
        </ApplyTheme>
      </View>
    </List.Item>
  )
}

UnassessedAssignment.propTypes = {
  assignment: object.isRequired // eslint-disable-line react/forbid-prop-types
}

export default UnassessedAssignment
