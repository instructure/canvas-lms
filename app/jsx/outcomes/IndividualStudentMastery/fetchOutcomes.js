/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import _ from 'lodash'
import uuid from 'uuid'
import parseLinkHeader from 'parse-link-header'

const deepMerge = (lhs, rhs) => {
  if (lhs === undefined || lhs === null) {
    return rhs
  } else if (Array.isArray(lhs)) {
    return lhs.concat(rhs)
  } else if (typeof lhs === 'object') {
    return _.mergeWith(lhs, rhs, deepMerge)
  } else {
    return rhs
  }
}

const combine = (promiseOfJson1, promiseOfJson2) => (
  Promise.all([promiseOfJson1, promiseOfJson2])
    .then(([json1, json2]) => deepMerge(json1, json2))
)

const parse = (response) => (
  response.text()
    .then((text) => (JSON.parse(text.replace('while(1);', ''))))
)

export const fetchUrl = (url) => (
  fetch(url, {
    credentials: 'include'
  })
    .then((response) => {
      const linkHeader = response.headers.get('link')
      const next = linkHeader ? parseLinkHeader(linkHeader).next : null
      if (next) {
        return combine(parse(response), fetchUrl(next.url))
      } else {
        return parse(response)
      }
    })
)

const fetchOutcomes = (courseId, studentId) => {
  let outcomeGroups
  let outcomeLinks
  let outcomeRollups
  let outcomeResultsByOutcomeId
  let assignmentsByAssignmentId

  return Promise.all([
    fetchUrl(`/api/v1/courses/${courseId}/outcome_groups?per_page=100`),
    fetchUrl(`/api/v1/courses/${courseId}/outcome_group_links?outcome_style=full&per_page=100`),
    fetchUrl(`/api/v1/courses/${courseId}/outcome_rollups?user_ids[]=${studentId}&per_page=100`)
  ])
    .then(([groups, links, rollups]) => {
      outcomeGroups = groups
      outcomeLinks = links
      outcomeRollups = rollups
    })
    .then(() => (
      Promise.all(outcomeLinks.map((outcomeLink) => (
        fetchUrl(`/api/v1/courses/${courseId}/outcome_results?user_ids[]=${studentId}&outcome_ids[]=${outcomeLink.outcome.id}&include[]=assignments&per_page=100`) // eslint-disable-line max-len
      )))
    ))
    .then((responses) => {
      outcomeResultsByOutcomeId = responses.reduce((acc, response, i) => {
        acc[outcomeLinks[i].outcome.id] = response.outcome_results.filter((r) => !r.hidden);
        return acc
      }, {})
      assignmentsByAssignmentId = _.keyBy(_.flatten(responses.map((response) => response.linked.assignments)), (a) => a.id)
    })
    .then(() => {
      const outcomes = outcomeLinks.map((outcomeLink) => ({
        // outcome ids are not unique (can appear in multiple groups), so we add unique
        // id to manage expansion/contraction
        expansionId: uuid(),
        groupId: outcomeLink.outcome_group.id,
        ...outcomeLink.outcome
      }))

      // filter empty outcome groups
      const outcomesByGroup = _.groupBy(outcomes, (o) => o.groupId)
      outcomeGroups = outcomeGroups.filter((g) => outcomesByGroup[g.id])

      // add rollup scores, mastered
      const outcomesById = _.keyBy(outcomes, (o) => o.id)
      outcomeRollups.rollups[0].scores.forEach((scoreData) => {
        const outcome = outcomesById[scoreData.links.outcome]
        if (outcome) {
          outcome.score = scoreData.score
          outcome.mastered = scoreData.score >= outcome.mastery_points
        }
      })

      // add results, assignments
      outcomes.forEach((outcome) => {
        outcome.results = outcomeResultsByOutcomeId[outcome.id] || [] // eslint-disable-line no-param-reassign
        outcome.results.forEach((result) => {
          result.assignment = assignmentsByAssignmentId[result.links.assignment] // eslint-disable-line no-param-reassign
        })
      })
      return { outcomeGroups, outcomes }
    })
}

export default fetchOutcomes
