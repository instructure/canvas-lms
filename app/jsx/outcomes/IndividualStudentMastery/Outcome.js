/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import _ from 'lodash'
import PropTypes from 'prop-types'
import I18n from 'i18n!IndividualStudentMasteryOutcome'
import {View, Flex} from '@instructure/ui-layout'
import {ToggleGroup} from '@instructure/ui-toggle-details'
import {List, Pill, Text, TruncateText} from '@instructure/ui-elements'
import natcompare from 'compiled/util/natcompare'
import AssignmentResult from './AssignmentResult'
import UnassessedAssignment from './UnassessedAssignment'
import OutcomePopover from './OutcomePopover'
import {ScreenReaderContent, PresentationContent} from '@instructure/ui-a11y'
import * as shapes from './shapes'

export default class Outcome extends React.Component {
  static propTypes = {
    outcome: shapes.outcomeShape.isRequired,
    expanded: PropTypes.bool.isRequired,
    onExpansionChange: PropTypes.func.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape
  }

  static defaultProps = {
    outcomeProficiency: null
  }

  handleToggle = (_event, expanded) => {
    this.props.onExpansionChange('outcome', this.props.outcome.expansionId, expanded)
  }

  renderHeader() {
    const {outcome, outcomeProficiency} = this.props
    const {assignments, display_name, mastered, title, score, points_possible, results} = outcome
    const numAlignments = assignments.length
    const pillAttributes = {margin: '0 0 0 x-small', text: I18n.t('Not mastered')}
    if (mastered) {
      Object.assign(pillAttributes, {text: I18n.t('Mastered'), variant: 'success'})
    }

    return (
      <Flex direction="row" justifyItems="space-between" data-selenium="outcome">
        <Flex.Item shrink>
          <Flex direction="column">
            <Flex.Item>
              <Text size="medium">
                <Flex>
                  <Flex.Item>
                    <OutcomePopover outcome={outcome} outcomeProficiency={outcomeProficiency} />
                  </Flex.Item>
                  <Flex.Item shrink padding="0 x-small">
                    <TruncateText>{display_name || title}</TruncateText>
                  </Flex.Item>
                </Flex>
              </Text>
            </Flex.Item>
            <Flex.Item>
              <Text size="small">
                {I18n.t(
                  {
                    zero: 'No alignments',
                    one: '%{count} alignment',
                    other: '%{count} alignments'
                  },
                  {count: I18n.n(numAlignments)}
                )}
              </Text>
            </Flex.Item>
          </Flex>
        </Flex.Item>
        <Flex.Item>
          {_.isNumber(score) && !_.every(results, ['hide_points', true]) && (
            <span>
              <PresentationContent>
                <Text size="medium">
                  {score}/{points_possible}
                </Text>
              </PresentationContent>
              <ScreenReaderContent>
                {I18n.t('%{score} out of %{points_possible} points', {score, points_possible})}
              </ScreenReaderContent>
            </span>
          )}
          <Pill {...pillAttributes} />
        </Flex.Item>
      </Flex>
    )
  }

  renderDetails() {
    const {outcome, outcomeProficiency} = this.props
    const {assignments, results} = outcome
    const assignmentsWithResults = _.filter(results, r =>
      r.assignment.id.startsWith('assignment_')
    ).map(r => r.assignment.id.split('_')[1])
    const assessmentsWithResults = _.filter(results, r =>
      r.assignment.id.startsWith('live_assessments/assessment_')
    ).map(r => r.assignment.id.split('_')[2])
    const unassessed = _.filter(
      assignments,
      a =>
        (a.assignment_id && !_.includes(assignmentsWithResults, a.assignment_id.toString())) ||
        (a.assessment_id && !_.includes(assessmentsWithResults, a.assessment_id.toString()))
    )
    return (
      <List variant="unstyled" delimiter="dashed">
        {results
          .sort(natcompare.byKey('submitted_or_assessed_at'))
          .reverse()
          .map(result => (
            <List.Item key={result.id}>
              <AssignmentResult
                result={result}
                outcome={outcome}
                outcomeProficiency={outcomeProficiency}
              />
            </List.Item>
          ))}
        {unassessed.map(assignment => (
          <UnassessedAssignment
            key={
              assignment.assessment_id ? `a${assignment.assessment_id}` : assignment.assignment_id
            }
            assignment={assignment}
          />
        ))}
      </List>
    )
  }

  renderEmpty() {
    return (
      <View as="div" padding="small">
        <Text>{I18n.t('No alignments are available for this outcome.')}</Text>
      </View>
    )
  }

  render() {
    const {outcome, expanded} = this.props
    const {assignments, title} = outcome
    const hasAlignments = assignments.length > 0
    return (
      <ToggleGroup
        summary={this.renderHeader()}
        toggleLabel={I18n.t('Toggle alignment details for %{title}', {title})}
        expanded={expanded}
        onToggle={this.handleToggle}
        border={false}
      >
        {hasAlignments ? this.renderDetails() : this.renderEmpty()}
      </ToggleGroup>
    )
  }
}
