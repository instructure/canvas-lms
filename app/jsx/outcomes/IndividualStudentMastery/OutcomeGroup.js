/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import I18n from 'i18n!IndividualStudentMasteryOutcomeGroup'
import {Flex, View} from '@instructure/ui-layout'
import {ToggleGroup} from '@instructure/ui-toggle-details'
import {List, Pill, Text} from '@instructure/ui-elements'
import natcompare from 'compiled/util/natcompare'
import Outcome from './Outcome'
import * as shapes from './shapes'

const outcomeGroupHeader = (outcomeGroup, numMastered, numGroup) => (
  <Flex justifyItems="space-between">
    <Flex.Item padding="0 x-small 0 0">
      <Text size="large" weight="bold">
        {outcomeGroup.title}
      </Text>
    </Flex.Item>
    <Flex.Item>
      <Pill text={I18n.t('%{numMastered} of %{numGroup} Mastered', {numMastered, numGroup})} />
    </Flex.Item>
  </Flex>
)

export default class OutcomeGroup extends React.Component {
  static propTypes = {
    outcomeGroup: shapes.outcomeGroupShape.isRequired,
    outcomes: PropTypes.arrayOf(shapes.outcomeShape).isRequired,
    expanded: PropTypes.bool.isRequired,
    expandedOutcomes: ImmutablePropTypes.set.isRequired,
    onExpansionChange: PropTypes.func.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape
  }

  static defaultProps = {
    outcomeProficiency: null
  }

  handleToggle = (_event, expanded) => {
    this.props.onExpansionChange('group', this.props.outcomeGroup.id, expanded)
  }

  render() {
    const {
      outcomeGroup,
      outcomes,
      expanded,
      expandedOutcomes,
      onExpansionChange,
      outcomeProficiency
    } = this.props
    const numMastered = outcomes.filter(o => o.mastered).length
    const numGroup = outcomes.length

    return (
      <View as="div" className="outcomeGroup">
        <ToggleGroup
          summary={outcomeGroupHeader(outcomeGroup, numMastered, numGroup)}
          toggleLabel={I18n.t('Toggle outcomes for %{title}', {title: outcomeGroup.title})}
          expanded={expanded}
          onToggle={this.handleToggle}
        >
          <List variant="unstyled" delimiter="solid">
            {outcomes.sort(natcompare.byKey('title')).map(outcome => (
              <List.Item key={outcome.id} margin="0">
                <Outcome
                  outcome={outcome}
                  expanded={expandedOutcomes.has(outcome.expansionId)}
                  onExpansionChange={onExpansionChange}
                  outcomeProficiency={outcomeProficiency}
                />
              </List.Item>
            ))}
          </List>
        </ToggleGroup>
      </View>
    )
  }
}
