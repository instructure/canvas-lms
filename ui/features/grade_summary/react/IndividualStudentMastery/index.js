/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import I18n from 'i18n!IndividualStudentMasteryIndex'
import {Flex} from '@instructure/ui-flex'
import {Text} from '@instructure/ui-text'
import {List} from '@instructure/ui-list'
import {Spinner} from '@instructure/ui-spinner'
import natcompare from '@canvas/util/natcompare'
import OutcomeGroup from './OutcomeGroup'
import fetchOutcomes from './fetchOutcomes'
import {Set} from 'immutable'
import * as shapes from './shapes'

class IndividualStudentMastery extends React.Component {
  static propTypes = {
    courseId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    studentId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    onExpansionChange: PropTypes.func,
    outcomeProficiency: shapes.outcomeProficiencyShape
  }

  static defaultProps = {
    onExpansionChange: () => {},
    outcomeProficiency: null
  }

  constructor() {
    super()
    this.state = {loading: true, error: null, expandedGroups: Set(), expandedOutcomes: Set()}
  }

  componentDidMount() {
    const {courseId, studentId} = this.props
    return fetchOutcomes(courseId, studentId)
      .then(({outcomeGroups, outcomes}) => {
        this.setState({outcomeGroups, outcomes})
      })
      .then(() => this.setState({loading: false}))
      .catch(e => this.setState({loading: false, error: e}))
  }

  onElementExpansionChange = (type, id, newState) => {
    let groups = this.state.expandedGroups
    let outcomes = this.state.expandedOutcomes
    if (type === 'group') {
      if (newState) {
        groups = groups.add(id)
      } else {
        groups = groups.delete(id)
        const idsToRemove = this.state.outcomes
          .filter(o => o.groupId === id)
          .map(o => o.expansionId)
        outcomes = outcomes.filterNot(oid => idsToRemove.includes(oid))
      }
    } else if (type === 'outcome') {
      if (newState) {
        outcomes = outcomes.add(id)
      } else {
        outcomes = outcomes.delete(id)
      }
    }
    this.setState(
      {
        expandedGroups: groups,
        expandedOutcomes: outcomes
      },
      () => this.notifyExpansionChange()
    )
  }

  contract() {
    this.setState(
      {
        expandedGroups: Set(),
        expandedOutcomes: Set()
      },
      () => this.notifyExpansionChange()
    )
  }

  expand() {
    this.setState(
      oldState => {
        const {outcomeGroups, outcomes} = oldState
        return {
          expandedGroups: Set(outcomeGroups.map(g => g.id)),
          expandedOutcomes: Set(outcomes.map(o => o.expansionId))
        }
      },
      () => this.notifyExpansionChange()
    )
  }

  notifyExpansionChange() {
    this.props.onExpansionChange(this.anyExpanded(), this.anyContracted())
  }

  anyExpanded() {
    return this.state.expandedGroups.size > 0 || this.state.expandedOutcomes.size > 0
  }

  anyContracted() {
    return (
      this.state.outcomeGroups.length > this.state.expandedGroups.size ||
      this.state.outcomes.length > this.state.expandedOutcomes.size
    )
  }

  renderLoading() {
    return (
      <Flex justifyItems="center" alignItems="center" padding="medium">
        <Flex.Item>
          <Spinner size="large" renderTitle={I18n.t('Loading outcome results')} />
        </Flex.Item>
      </Flex>
    )
  }

  renderError() {
    return (
      <Flex justifyItems="start" alignItems="center" padding="medium 0">
        <Flex.Item>
          <Text color="danger">{I18n.t('An error occurred loading outcomes data.')}</Text>
        </Flex.Item>
      </Flex>
    )
  }

  renderEmpty() {
    return (
      <Flex justifyItems="start" alignItems="center" padding="medium 0">
        <Flex.Item>
          <Text>{I18n.t('There are no outcomes in the course.')}</Text>
        </Flex.Item>
      </Flex>
    )
  }

  renderGroups() {
    const {outcomeGroups, outcomes} = this.state
    const {outcomeProficiency} = this.props
    return (
      <div>
        <List variant="unstyled">
          {outcomeGroups.sort(natcompare.byKey('title')).map(outcomeGroup => (
            <List.Item key={outcomeGroup.id}>
              <OutcomeGroup
                outcomeGroup={outcomeGroup}
                outcomes={outcomes.filter(o => o.groupId.toString() === outcomeGroup.id.toString())}
                expanded={this.state.expandedGroups.has(outcomeGroup.id)}
                expandedOutcomes={this.state.expandedOutcomes}
                onExpansionChange={this.onElementExpansionChange}
                outcomeProficiency={outcomeProficiency}
              />
            </List.Item>
          ))}
        </List>
      </div>
    )
  }

  render() {
    const {error, loading, outcomeGroups} = this.state

    if (loading) {
      return this.renderLoading()
    } else if (error) {
      return this.renderError()
    } else if (outcomeGroups.length === 0) {
      return this.renderEmpty()
    } else {
      return this.renderGroups()
    }
  }
}

export default IndividualStudentMastery
