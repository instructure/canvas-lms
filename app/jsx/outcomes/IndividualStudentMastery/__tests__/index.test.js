/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {mount, shallow} from 'enzyme'
import {Set} from 'immutable'
import IndividualStudentMastery from '../index'
import fetchOutcomes from '../fetchOutcomes'

jest.mock('../fetchOutcomes')

beforeEach(() => {
  fetchOutcomes.mockImplementation(() => Promise.resolve(null))
})

const props = {
  studentId: 12,
  courseId: 110,
  onExpansionChange: jest.fn()
}

it('renders the component', () => {
  const wrapper = shallow(<IndividualStudentMastery {...props} />)
  expect(wrapper).toMatchSnapshot()
})

it('attempts to load when mounted', () => {
  mount(<IndividualStudentMastery {...props} />)
  expect(fetchOutcomes).toHaveBeenCalled()
})

it('renders loading before promise resolves', () => {
  fetchOutcomes.mockImplementation(() => new Promise(() => {})) // unresolved promise
  const wrapper = mount(<IndividualStudentMastery {...props} />)
  expect(wrapper.find('Spinner')).toHaveLength(1)
})

it('renders error when error occurs during fetch', async () => {
  fetchOutcomes.mockImplementation(() => Promise.reject(new Error('foo')))
  const wrapper = mount(<IndividualStudentMastery {...props} />)
  await wrapper.instance().componentDidMount()
  expect(wrapper.text()).toMatch('An error occurred')
})

it('renders empty if no groups are returned', async () => {
  fetchOutcomes.mockImplementation(() => Promise.resolve({outcomeGroups: [], outcomes: []}))
  const wrapper = mount(<IndividualStudentMastery {...props} />)
  await wrapper.instance().componentDidMount()
  expect(wrapper.text()).toMatch('There are no outcomes in the course')
})

it('renders outcome groups if they are returned', async () => {
  fetchOutcomes.mockImplementation(() =>
    Promise.resolve({
      outcomeGroups: [{id: 1, title: 'Group'}],
      outcomes: []
    })
  )
  const wrapper = shallow(<IndividualStudentMastery {...props} />)
  await wrapper.instance().componentDidMount()
  expect(wrapper.update().find('OutcomeGroup')).toHaveLength(1)
})

describe('expand and contract', () => {
  beforeEach(() => {
    fetchOutcomes.mockImplementation(() =>
      Promise.resolve({
        outcomeGroups: [{id: 1, title: 'Group'}],
        outcomes: [
          {
            id: 2,
            expansionId: 100,
            groupId: 1,
            title: 'Outcome',
            mastered: false,
            mastery_points: 0,
            points_possible: 0,
            calculation_method: 'highest',
            assignments: [],
            results: [],
            ratings: []
          }
        ]
      })
    )
  })

  it('toggles elements to expanded when event fired', async () => {
    const wrapper = mount(<IndividualStudentMastery {...props} />)
    await wrapper.instance().componentDidMount()
    wrapper.instance().onElementExpansionChange('outcome', 100, true)
    expect(wrapper.state('expandedOutcomes').equals(Set.of(100))).toBe(true)
    wrapper.instance().onElementExpansionChange('outcome', 100, false)
    expect(wrapper.state('expandedOutcomes').equals(Set())).toBe(true)
  })

  it('contracts child outcomes when a group is contracted', async () => {
    const wrapper = mount(<IndividualStudentMastery {...props} />)
    await wrapper.instance().componentDidMount()
    wrapper.setState({expandedGroups: Set.of(1), expandedOutcomes: Set.of(100)}, () => {
      wrapper.instance().onElementExpansionChange('group', 1, false)
    })
    expect(wrapper.state('expandedOutcomes').equals(Set())).toBe(true)
  })

  it('expands all when expand() is called', async () => {
    const wrapper = mount(<IndividualStudentMastery {...props} />)
    await wrapper.instance().componentDidMount()
    wrapper.instance().expand()
    expect(wrapper.state('expandedGroups').equals(Set.of(1))).toBe(true)
    expect(wrapper.state('expandedOutcomes').equals(Set.of(100))).toBe(true)
  })

  it('contracts all when contract() is called', async () => {
    const wrapper = mount(<IndividualStudentMastery {...props} />)
    await wrapper.instance().componentDidMount()
    wrapper.setState({expandedOutcomes: Set.of(100), expandedGroups: Set.of(1)}, () => {
      wrapper.instance().contract()
    })
    expect(wrapper.state('expandedGroups').equals(Set())).toBe(true)
    expect(wrapper.state('expandedOutcomes').equals(Set())).toBe(true)
  })

  it('notifies when expansion is changed', async () => {
    const wrapper = mount(<IndividualStudentMastery {...props} />)
    await wrapper.instance().componentDidMount()
    wrapper.instance().onElementExpansionChange('outcome', 100, true)
    expect(props.onExpansionChange).toHaveBeenLastCalledWith(true, true)
    wrapper.instance().onElementExpansionChange('group', 1, true)
    expect(props.onExpansionChange).toHaveBeenLastCalledWith(true, false)
    wrapper.instance().contract()
    expect(props.onExpansionChange).toHaveBeenLastCalledWith(false, true)
  })
})

it('renders outcome groups in alphabetical order by title', async () => {
  fetchOutcomes.mockImplementation(() =>
    Promise.resolve({
      outcomeGroups: [
        {id: 1, title: 'ZZ Top Albums'},
        {id: 2, title: 'Aerosmith Albums'},
        {id: 3, title: 'Aardvark Albums'},
        {id: 4, title: 'abba Albums'}
      ],
      outcomes: []
    })
  )
  const wrapper = shallow(<IndividualStudentMastery {...props} />)
  await wrapper.instance().componentDidMount()
  const groups = wrapper.find('OutcomeGroup')
  expect(groups).toHaveLength(4)
  expect(groups.get(0).props.outcomeGroup.title).toEqual('Aardvark Albums')
  expect(groups.get(1).props.outcomeGroup.title).toEqual('abba Albums')
  expect(groups.get(2).props.outcomeGroup.title).toEqual('Aerosmith Albums')
  expect(groups.get(3).props.outcomeGroup.title).toEqual('ZZ Top Albums')
})
