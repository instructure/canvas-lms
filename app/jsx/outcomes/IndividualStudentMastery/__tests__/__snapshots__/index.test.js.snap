// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<Flex
  alignItems="center"
  as="span"
  direction="row"
  elementRef={[Function]}
  inline={false}
  justifyItems="center"
  padding="medium"
  visualDebug={false}
  wrapItems={false}
>
  <FlexItem
    as="span"
    elementRef={[Function]}
    grow={false}
    shrink={false}
  >
    <Spinner
      as="div"
      renderTitle="Loading outcome results"
      size="large"
      variant="default"
    />
  </FlexItem>
</Flex>
`;
